gazelle/pythonconfig/pythonconfig.go

```diff
@@ -443,3 +460,46 @@ func (c *Config) SetTestFilePattern(patterns []string) {
 func (c *Config) TestFilePattern() []string {
 	return c.testFilePattern
 }
+
+// SetLabelConvention sets the label convention used for third-party dependencies.
+func (c *Config) SetLabelConvention(convention string) {
+	c.labelConvention = convention
+}
+
+// LabelConvention returns the label convention used for third-party dependencies.
+func (c *Config) LabelConvention() string {
+	return c.labelConvention
+}
+
+// SetLabelConvention sets the label normalization applied to distribution names of third-party dependencies.
+func (c *Config) SetLabelNormalization(normalizationType LabelNormalizationType) {
+	c.labelNormalization = normalizationType
+}
+
+// LabelConvention returns the label normalization applied to distribution names of third-party dependencies.
+func (c *Config) LabelNormalization() LabelNormalizationType {
+	return c.labelNormalization
+}
+
+// FormatThirdPartyDependency returns a label to a third-party dependency performing all formating and normalization.
+func (c *Config) FormatThirdPartyDependency(repositoryName string, distributionName string) label.Label {
+	conventionalDistributionName := strings.ReplaceAll(c.labelConvention, distributionNameLabelConventionSubstitution, distributionName)
+
+	var normConventionalDistributionName string
+	switch norm := c.LabelNormalization(); norm {
+	case SnakeCaseLabelNormalizationType:
+		normConventionalDistributionName = strings.ToLower(conventionalDistributionName)
+		normConventionalDistributionName = strings.ReplaceAll(normConventionalDistributionName, "-", "_")
+		normConventionalDistributionName = strings.ReplaceAll(normConventionalDistributionName, ".", "_")
```

Comment:
we also need to replace any occurrences of `__` to `_`. This is done in our starlark code, so it is essentially the same as `pep503` but with `_`, so doing the same regex as below would be ideal.

Reply:
This piece was moved exactly from `SanitizeDistribution` to make sure nothing breaks existing behaviour. Could you please point to the Starlark code you had in mind? 

I found [this](https://github.com/bazelbuild/rules_python/blob/49d180fc773da54df63386c64790dbf8b7616735/python/private/py_wheel.bzl#L229-L264), and from what my internal 🧠 interpreter tells me it escapes char by char, so `distribution----with--dashes` becomes `distribution____with__dashes`, not `distribution_with_dashes` if I replace `"[-_.]+"` with `_` as in `pep503`.

Reply:
The rules_python//python/private:normalize_name.bzl is the code that I had in mind.

Reply:
Thank you. I updated the code and actually found extra corner case where `normalize_name.bz` behaves differently than `SnakeCaseLabelNormalizationType` and the old code: prefix/trailing dashes. This has been fixed, the mapping updated, more tests added.